cdap-app-fabric/src/main/java/io/cdap/cdap/app/deploy/ProgramRunDispatcher.java

```diff
@@ -0,0 +1,36 @@
+/*
+ * Copyright © 2022 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.cdap.app.deploy;
+
+import io.cdap.cdap.app.runtime.ProgramController;
+import io.cdap.cdap.internal.app.deploy.pipeline.ProgramRunInfo;
+
+/**
+ * This interface is used for providing the abstraction for executing program-run logic either
+ * In-memory or remotely (e.g., in a system pod)
+ */
+public interface ProgramRunDispatcher {
+
+  /**
+   *
+   * @param programRunInfo Context Information required to perform program-run operation.
+   * @return {@link ProgramController} Controller containing state for the program.
+   * @throws Exception
+   */
+  ProgramController dispatchProgram(ProgramRunInfo programRunInfo) throws Exception;
```

Comment:
are there any specific exceptions that are expected to be thrown while dispatching a program run? If so, we should document them in the javadoc

Reply:
Have put generic exception in definition to make sure we don't miss out on any exception. Will fine-tune this later if requried.